apps/api/src/queries/digital-text/digital-text-query.service.ts

```diff
@@ -0,0 +1,82 @@
+import { isDeepStrictEqual } from 'util';
+import { CoscradUserWithGroups } from '../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
+import { AggregateCompositeIdentifier } from '../../domain/types/AggregateCompositeIdentifier';
+import { AggregateId } from '../../domain/types/AggregateId';
+import { AggregateType } from '../../domain/types/AggregateType';
+import { Maybe } from '../../lib/types/maybe';
+import { NotFound } from '../../lib/types/not-found';
+import cloneToPlainObject from '../../lib/utilities/cloneToPlainObject';
+import { ArangoEventRepository } from '../../persistence/repositories/arango-event-repository';
+import { DigitalTextViewModel } from './digital-text.view-model';
+
+export class DigitalTextQueryService {
+    /**
+     * TODO We probably want to make this depend on a `DigitalTextQueryRepository`
+     * that itself depends on the event repository. This is because we will eventually
+     * publish events from the domain onto a messaging queue and cache the event-sourced
+     * query models in a second database, achieving full CQRS-ES (big optimization).
+     */
+    constructor(private readonly eventRepository: ArangoEventRepository) {}
+
+    public async fetchById(
+        id: string,
+        userWithGroups?: CoscradUserWithGroups
+    ): Promise<Maybe<DigitalTextViewModel>> {
+        const fullEventHistory = await this.eventRepository.fetchEvents();
+
+        const eventHistoryForThisDigitalText = fullEventHistory.filter(
+            ({ payload: { aggregateCompositeIdentifier } }) =>
+                isDeepStrictEqual(aggregateCompositeIdentifier, this.buildCompositeIdentifier(id))
+        );
+
+        if (eventHistoryForThisDigitalText.length === 0) return NotFound;
+
+        /**
+         * Note that we require the entire event history in order to denormalize
+         * the query models. We may want to determine things like
+         * - which tags apply to this resource
+         * - which other resources is this resource connected to
+         * - which categories apply to this resource
+         */
+        const hydratedViewModel = new DigitalTextViewModel(id).applyStream(fullEventHistory);
+
+        if (hydratedViewModel.isPublished || hydratedViewModel.hasReadAccess(userWithGroups))
+            // TODO should we remove any fields?
+            return cloneToPlainObject(hydratedViewModel);
+
+        return NotFound;
```

Comment:
This seems like the reverse of how we normally return data from a function?  Wouldn't you want to return early with the `NotFound` message on the negation of `.isPublished` and `.hasReadAccess` and have the final return be the `cloneToPlainObject(hydratedViewModel)`?  Obviously just a detail but I'm curious

Reply:
This is essentially a ternary (except I use return earlier as I find that a bit more readable). In this situation, there are 2 reasons why I did it in this order:
 1. I prefer to have the more complex logic first in a ternary as it is easier to parse that immediately as a reader and "hold that thought" for the simple default case.
 2. I try to prefer the positive form of a logical expression when possible